File: src/www/__test__/interpreter/FileInterpreter.js
import chalk from "chalk";
import Context from "./Context";
import FileTokenizer from "./FileTokenizer";
import TestInterpreter from "./TestInterpreter";
import InstructionSequence from "./InstructionSequence";
import decodeContent from "../macros/decodeContent";

export default class FileInterpreter {
  _befores = [];
  _afters = [];
  _rootContext = new Context();

  afterTest(fn) {
    this._afters.push(fn);
    return this;
  }

  beforeTest(fn) {
    this._befores.push(fn);
    return this;
  }

  configure(fn) {
    fn(this._rootContext);
    return this;
  }

  interpretFile(file) {
    const fileTokenizer = new FileTokenizer(file);

    try {
      this._interpretFile(fileTokenizer);
    } catch (e) {
      console.error(`Error at:\n${fileTokenizer.toString()}`);
      throw e;
    }
  }

  _skipUntilTest(fileTokenizer) {
    fileTokenizer.skipUntil(/^##[^#]/i);
  }

  _interpretFile(fileTokenizer) {
    this._skipUntilTest(fileTokenizer);
    while (!fileTokenizer.eof()) {
      this._interpretTest(fileTokenizer);
    }
  }

  _interpretTest(fileTokenizer) {
    const title = fileTokenizer.accept();
    const lineNumber = fileTokenizer.getLineNumber();
    const content = fileTokenizer.acceptUntil(/^##[^#]/).join("\n");

    const testContext = this._rootContext.clone();
    const testInterpreter = new TestInterpreter(testContext);
    const instructions = decodeContent("", lineNumber, content);
    const instructionSequence = new InstructionSequence(instructions);

    test(title, async () => {
      try {
        this._befores.forEach(fn => fn(testContext));
        await testInterpreter.interpretTest(instructionSequence);
        this._afters.forEach(fn => fn(testContext));
      } catch (e) {
        e.message +=
          "\n\n" +
          chalk.bold(title) +
          "\n" +
          instructionSequence.toString() +
          "\n";

        throw e;
      }
    });
  }
}

File: src/test/java/com/drpicox/game/interpreter/FileInterpreter.java
package com.drpicox.game.interpreter;

import com.drpicox.game.macros.AllMacros;
import org.junit.jupiter.api.DynamicTest;
import org.springframework.stereotype.Component;

import java.util.LinkedList;
import java.util.List;

@Component
public class FileInterpreter {

    private AllMacros allMacros;
    private TestInterpreter testInterpreter;

    public FileInterpreter(AllMacros allMacros, TestInterpreter testInterpreter) {
        this.allMacros = allMacros;
        this.testInterpreter = testInterpreter;
    }

    public List<DynamicTest> interpretFile(String file, Context context) {
        var tokens = new FileTokenizer(file);

        return this.interpretFile(tokens, context);
    }

    private List<DynamicTest> interpretFile(FileTokenizer tokens, Context context) {
        var tests = new LinkedList<DynamicTest>();

        this.skipUntilTest(tokens);
        while (!tokens.eof()) {
            var test = this.interpretTest(tokens, context);
            tests.add(test);
        }

        return tests;
    }

    private void skipUntilTest(FileTokenizer tokens) {
        tokens.skipUntil("##[^#].*");
    }

    private InstructionSequence acceptTestInstructions(FileTokenizer tokens) {
        var lineNumber = tokens.getLineNumber();
        var instructionsSources = tokens.acceptUntil("##[^#].*");

        var instructions = this.allMacros.decodeInstructions("", lineNumber, instructionsSources);
        return new InstructionSequence(instructions);
    }

    private DynamicTest interpretTest(FileTokenizer tokens, Context context) {
        var title = tokens.accept();
        var testContext = context.clone();
        var instructionSequence = this.acceptTestInstructions(tokens);
        return testInterpreter.interpretTest(title, testContext, instructionSequence);
    }
}
